### PR TITLE
Add Restructure

### PR DESCRIPTION
**Pull request by gvcallen**
Adds the Restructure bijector similar to [tfb.Restructure](https://www.tensorflow.org/probability/api_docs/python/tfp/bijectors/Restructure).

@lockwo I'm not sure how you feel about adding these types of bijectors into distreqx, however I feel they are very necessary if the end goal is a self-contained package with no interop with Tensorflow. Although I'm not sure which direction you are intending for the library to go, I feel there's a definite need in the Equinox ecosystem for a native distribution + bijector library, which these kinds of classes are necessary for.

As an example "critical" use case for this `Restructure` bijector, I have found it is essentially impossible to easily model correlations between arbitrary pytree leaves without it, without maintaining lots of boilerplate. In my specific case, I am training a normalizing flow on samples of specific leaves in my tree. `Restructure` can then be used (along with `Split`) to restructure the flow's output so that it properly represents a distribution over my specific tree, as opposed to only a sub-portion of my tree's leaves flattened which then I must map manually. Combined with a `Joint` distribution and potentially a multi-bijector similar to tbf's `JointMap`, and I think its then possible to model arbitrary correlations across arbitrary pytree structures and leaves.

### CODE REVIEW

**Comment by lockwo** [comment]
Yea I think these are in general ok add. Although distrax doesn't have them, I think distrax was very array focused. I think most users (including myself) generally work with arrays, and this pytree stuff is strictly additive so users concerned with only arrays are ok either way.

**Comment by gvcallen** [comment]
@lockwo ah I see, yes I'm personally dealing with very nested Pytrees all the time :D

File: distreqx/bijectors/_restructure.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+    in_treedef: PyTreeDef = eqx.field(static=True)  # type: ignore
+    out_treedef: PyTreeDef = eqx.field(static=True)  # type: ignore
+    forward_permutation: tuple[int, ...] = eqx.field(static=True)
+    inverse_permutation: tuple[int, ...] = eqx.field(static=True)
+
+    _is_constant_jacobian: bool = True
+    _is_constant_log_det: bool = True
+
+    def __init__(self, in_structure: Any, out_structure: Any):
```
Comment (lockwo) [inline distreqx/bijectors/_restructure.py:36]: we can type as PyTree rather than Any?

File: distreqx/bijectors/_restructure.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        # Pre-compute the routing permutations for fast forward/inverse passes
+        self.forward_permutation = tuple(flat_in.index(token) for token in flat_out)
+        self.inverse_permutation = tuple(flat_out.index(token) for token in flat_in)
+
+    def forward_and_log_det(self, x: Any) -> tuple[Any, Array]:
+        """Computes y = restructure(x) and log|det J(f)(x)| = 0."""
+        if jax.tree_util.tree_structure(x) != self.in_treedef:
+            raise ValueError("Input `x` does not match the expected `in_structure`.")
+
+        flat_x = jax.tree_util.tree_leaves(x)
```
Comment (lockwo) [inline distreqx/bijectors/_restructure.py:75]: jax.tree.leaves
